### PR TITLE
Assert no side effects occur on test

### PR DESCRIPTION
**Pull request by farazdagi**
**What type of PR is this?**
Bug fix

**What does this PR do? Why is it needed?**
- Due to inaccurate params overriding, we had our test functions producing side effects, and thus leaking the state out of tests. This PR fixes that.

**Which issues(s) does this PR fix?**
- Issue located and partial fix provided in #5691 

**Other notes for review**
- In number of places instead of `init()`, test setup code has been moved to `TestMain` (which has exact purpose of being function to setup/teardown common state before tests). `TestMain` was created in `packagename_test.go` file - so it will be easier to find it (fishing for duplicated `init()` funcs is no fun)
- Changes to existing code are minimal: 
  - Whenever one wants to modify configuration parameters within a **test case** `params.SetupTestConfigCleanup(t *testing.T)` needs to be called. Inside that function [t.Cleanup](https://tip.golang.org/pkg/testing/#T.Cleanup) is used, and once all test cases's tests are done - cleanup resets the state. So, within test case - any modifications to existing config can be done w/o limitations (as they will be reset when test case is cleaned up).
  - Whenever one want to modify state for a whole **test suite** (all tests located w/i a given package), `TestMain` is to be created
```golang
func TestMain(m *testing.M) {
	// Use minimal config to reduce test setup time.
	prevConfig := params.BeaconConfig().Copy()
	params.OverrideBeaconConfig(params.MinimalSpecConfig())

	retVal := m.Run()

	// Reset configuration.
	params.OverrideBeaconConfig(prevConfig)
	os.Exit(retVal)
}
```

### CODE REVIEW

File: beacon-chain/core/blocks/eth1_data_test.go [hunk 1]
```diff
@@ -58,9 +58,10 @@ func TestEth1DataHasEnoughSupport(t *testing.T) {
 
 	for i, tt := range tests {
 		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
-			c := params.BeaconConfig()
+			c := params.BeaconConfig().Copy()
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/eth1_data_test.go:61]: Where did this Copy method come from? Seems like a very easy way to misuse this API.

Comment (farazdagi) [reply beacon-chain/core/blocks/eth1_data_test.go:61]: I've added that method in #5691

I needed a method to copy state, so that we do not operate on the same pointer data. API, as used now, is: get pointer, update some value (changing the original object), override. I wanted to have a way to manipulate the params w/o side effects. So, I operate on copy of the object, when overriding I get a resetting function - so I can get back to the exact original state.

Thus, my understanding is that having a global state, which has its values mutated inside a unit test, is not a safe way to use the API. Hence, the proposed solution. Do you have any suggestions on how this should be done, instead?

Comment (prestonvanloon) [reply beacon-chain/core/blocks/eth1_data_test.go:61]: I have two ideas:
- Full copy on read
- Keep the original copy in memory, copy-on-write for reset.

I think that using Copy() will be difficult to enforce unless we add another static analysis check and that it won't be very clear to users that they should use the copy method in tests. I wish we had immutable objects in go :(

Comment (farazdagi) [reply beacon-chain/core/blocks/eth1_data_test.go:61]: I suppose that full copy on read (in `params.BeaconChain()`, `params.MainnetConfig()`, and `params.MinimalSpecConfig()`) is the most straight-forward way to proceed.

So, I will make sure that `params.BeaconConfig()` returns copy of the object, and have the sth like this:
```go
// Updating some params
config := params.BeaconConfig()
config.foo = "bar" // safe to do on copy
resetCfg := params.OverrideBeaconConfigWithReset(config)
defer resetCfg()

// Using certain preset config
resetCfg := params.OverrideBeaconConfigWithReset(params.MainnetConfig())
defer resetCfg()
```

I will also reduce scope of `params.copy()` to package-level.
Thoughts?

Comment (prestonvanloon) [reply beacon-chain/core/blocks/eth1_data_test.go:61]: The side effect of copy on read is that there will be much more memory churn in the production code path. Maybe we can try the other idea. Something like

originalDefaultConfig := deepCopy(defaultConfig)

func reset() {
  defaultConfig = deepCopy(originalDefaultConfig)
}

Comment (prestonvanloon) [reply beacon-chain/core/blocks/eth1_data_test.go:61]: The side effect of copy on read is that there will be much more memory churn in the production code path. Maybe we can try the other idea. Something like

originalDefaultConfig := deepCopy(defaultConfig)

func reset() {
  defaultConfig = deepCopy(originalDefaultConfig)
}

Comment (farazdagi) [reply beacon-chain/core/blocks/eth1_data_test.go]: So, keeping in mind that we want to:
- continue returning pointers on read, in order not to introduce more memory churn
- be able to modify config params individually and as a whole (that's changing mainnet/miniconf)
- make sure that once test case is done, teardown code is invoked, which restores previous global state - i.e. makes sure that there's no side-effects leaked out of test case

I've updated code to use testing.T.Cleanup, which is available as of Go 1.14, and allows to register cleanup function with the test case. So, whenever, within a testcase params modification are needed, one needs to invoke `params.SetupTestConfigCleanup(t *testing.T)`, which does the following:
```golang
// SetupTestConfigCleanup preserves configurations allowing to modify them 
// within tests without any restrictions, everything is restored after the test.
func SetupTestConfigCleanup(t *testing.T) {
	prevDefaultBeaconConfig := defaultBeaconConfig.Copy()
	prevBeaconConfig := beaconConfig.Copy()
	t.Cleanup(func() {
		defaultBeaconConfig = prevDefaultBeaconConfig
		beaconConfig = prevBeaconConfig
	})
}
```

Now, within individual tests not much is changing, for example:
```golang
func TestDiscv5_AddRetrieveForkEntryENR(t *testing.T) {
	params.SetupTestConfigCleanup(t) // Configure teardown function
	c := params.BeaconConfig() // Get pointer and modify values as needed
	c.ForkVersionSchedule = map[uint64][]byte{
		0: params.BeaconConfig().GenesisForkVersion,
		1: {0, 0, 0, 1},
	}
	nextForkEpoch := uint64(1)
	nextForkVersion := []byte{0, 0, 0, 1}
	c.NextForkEpoch = nextForkEpoch
	c.NextForkVersion = nextForkVersion
	params.OverrideBeaconConfig(c)
        // ...
        // Once test is finished, cleanup code is called.
}
```

File: beacon-chain/sync/subscriber_test.go [hunk 2]
```diff
@@ -92,6 +92,7 @@ func TestSubscribe_ReceivesAttesterSlashing(t *testing.T) {
 	topic := "/eth2/%x/attester_slashing"
 	var wg sync.WaitGroup
 	wg.Add(1)
+	params.SetupTestConfigCleanup(t)
```
Comment (prestonvanloon) [inline beacon-chain/sync/subscriber_test.go:95]: I think this is still going to be fragile. How will we ensure users always clean up the config?

Comment (farazdagi) [reply beacon-chain/sync/subscriber_test.go:95]: Whenever using a resource, you either created it or get reference for it. If you create a resource (say database to be used in test), then you need to tear it down. In #5725 I made sure that when db resource is created, teardown is automatically attached. There weren't any checks to ensure users clean up - it was done manually.

In case of configuration which is global variable to which we want to keep accessing by pointer, then there's no other option but to register cleanup function (and you need to register it **before** you modify pointed data). The `SetupTestConfigCleanup()` is exactly that - registering teardown function. At the moment we do nothing to teardown - we just modify the configuration object, with `SetupTestConfigCleanup()` we have a way to attach cleanup routine - all the rest of the code remains the same.

To sum up: when setting up resource, teardown is responsibility of users, especially when it comes to *writing tests*. We do not ensure it in any way (and I am not sure how we are supposed to do so, except for some static analysis tool). We are using global var, pointer to that global var, and if we continue to do so - then I am not sure what else can be done.

Currently, we are leaking the state - and we cannot avoid it. The `SetupTestConfigCleanup()` function allows you to restore the state.

**Comment by nisdas** [comment]
@farazdagi please fix conflicts
